### ISSUE TITLE
Inability to connect from clients preferring group exchange when moduli file is not found/loaded

### ISSUE THREAD

**Issue by mdietzer-fn** [paramiko/paramiko#2126]
Currently, there is an issue where if a client prefers a group exchange method, but the server has no modulus file, the server will advertise supporting it anyway and then the negotiation will (expectedly) fail.

https://github.com/paramiko/paramiko/blob/main/paramiko/transport.py#L2316-L2328

The problem is here in the first line, a list copy is made from the supported KEX algorithms, however after that, the transport's supported KEX algorithms list is updated directly, not the copy.
Therefor, the server will still advertise support (as it sends the data from its copy), despite not actually having support for these algorithms.

Either the entire operation should be performed on the (copied) list, or the transport's state.

This will yield weird errors, such as `paramiko.ssh_exception.SSHException: Expecting packet from (30,), got 34` for the affected clients.

**Comment by bskinn** [comment]
Flagging this to include/consider in the key/auth rewrite of #387.

Thanks for the report, @mdietzer-fn!

**Comment by qinwang-murphy** [comment]
@mdietzer-fn  hi,when do we plan to fix the problem?

**Comment by jun66j5** [comment]
It seems that #2083 fixes this issue. Please try it.

**Comment by fathom-parth** [comment]
Is there a workaround here if you don't control the client? Creating a moduli file?

**Comment by mdietzer-fn** [comment]
> Is there a workaround here if you don't control the client? Creating a moduli file?

Yep. You can use one from an existing Linux distro or generate your own (this will take some time)

**Comment by Route-Reflector** [comment]
### Re-verified on paramiko 5.0.0: bug still present

Sharing observations from a downstream project — I hit this exact bug while using paramiko's server mode against netmiko clients like FortinetSSH, and re-verified the behavior on the recently released paramiko 5.0.0 (2026-05-09).

**TL;DR:**
- The root cause described in this issue is **still present in paramiko 5.0.0**.
- The failure mode depends on the client's kex preference list (table below), but it is still a broken connection from the user's perspective.
- The fix proposed in #2083 still applies and is still correct.

#### Failure modes on paramiko 5.0.0

The stale-snapshot bug in [`_send_kex_init()` (transport.py L2257)](https://github.com/paramiko/paramiko/blob/5.0.0/paramiko/transport.py#L2257) is unchanged. What did change is that [`_parse_kex_init()` (transport.py L2417)](https://github.com/paramiko/paramiko/blob/5.0.0/paramiko/transport.py#L2417) now intersects against the runtime `self.preferred_kex` (post-filter):

```python
if self.server_mode:
    agreed_kex = list(filter(self.preferred_kex.__contains__, kex_algo_list))
else:
    agreed_kex = list(filter(kex_algo_list.__contains__, self.preferred_kex))
if len(agreed_kex) == 0:
    raise IncompatiblePeer("Incompatible ssh peer (no acceptable kex algorithm)")
```

The server's KEXINIT (built from the stale `kex_algos` snapshot) still advertises GEX, while the server's matching uses the post-filter `preferred_kex` (no GEX). The resulting failure depends on what the client offers back:

| Client kex preference | paramiko 4.x | paramiko 5.0.0 |
| --- | --- | --- |
| GEX only (extreme case) | `MessageOrderError` | `IncompatiblePeer` (intersection empty) |
| GEX + non-GEX fallback | `MessageOrderError` | `MessageOrderError` (server picks fallback, client picks GEX, packet types mismatch) |

The second row is the common real-world case (e.g. netmiko's `FortinetSSH`, or `ssh -oKexAlgorithms=diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha256`). `IncompatiblePeer` is paramiko-5.0-specific and only surfaces when the client offers GEX exclusively.

Either way, the server ends up self-inconsistent (advertises GEX, then doesn't accept GEX) and the connection breaks.

<details>
<summary>Minimal repro on paramiko 5.0.0 (paramiko-only, no external SSH client needed)</summary>

```python
import socket
import threading
import paramiko

class _Server(paramiko.ServerInterface):
    def check_auth_password(self, u, p):
        return paramiko.AUTH_SUCCESSFUL
    def check_channel_request(self, kind, _):
        return (paramiko.OPEN_SUCCEEDED if kind == "session"
                else paramiko.OPEN_FAILED_ADMINISTRATIVELY_PROHIBITED)
    def get_allowed_auths(self, _):
        return "password"

def _server_thread(sock, host_key, result):
    client_sock, _ = sock.accept()
    session = paramiko.Transport(client_sock)
    # NOTE: deliberately NOT calling load_server_moduli()
    # NOTE: deliberately NOT setting disabled_algorithms (no workaround)
    session.add_server_key(host_key)
    try:
        session.start_server(server=_Server())
        result["server"] = "ok"
    except Exception as exc:
        result["server"] = f"{type(exc).__name__}: {exc}"
    finally:
        session.close()

host_key = paramiko.RSAKey.generate(2048)
sock = socket.socket()
sock.bind(("127.0.0.1", 0))
sock.listen(1)
port = sock.getsockname()[1]
result = {}
server_t = threading.Thread(
    target=_server_thread, args=(sock, host_key, result), daemon=True,
)
server_t.start()

client = paramiko.SSHClient()
client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
try:
    client.connect(
        "127.0.0.1", port=port, username="u", password="p",
        # Force the client to prefer GEX only — common real case is GEX + fallback
        # (e.g. FortinetSSH); GEX-only here surfaces the 5.0-specific IncompatiblePeer path.
        disabled_algorithms={"kex": [
            "curve25519-sha256", "curve25519-sha256@libssh.org",
            "ecdh-sha2-nistp256", "ecdh-sha2-nistp384", "ecdh-sha2-nistp521",
            "diffie-hellman-group16-sha512", "diffie-hellman-group14-sha256",
            "diffie-hellman-group14-sha1", "diffie-hellman-group1-sha1",
        ]},
        allow_agent=False, look_for_keys=False,
        timeout=5, banner_timeout=5, auth_timeout=5,
    )
    print("client: ok")
except Exception as exc:
    print(f"client: {type(exc).__name__}: {exc}")
server_t.join(timeout=2)
print(f"server: {result.get('server')}")
```

Output on paramiko 5.0.0 / Python 3.13 (server-side error is the actual indicator; client-side exception type can vary with socket teardown timing):

```
client: <EOFError | ConnectionResetError | SSHException>
server: IncompatiblePeer: Incompatible ssh peer (no acceptable kex algorithm)
```

Swapping the client's `disabled_algorithms` to allow a non-GEX fallback (e.g. drop `diffie-hellman-group14-sha256` from the disabled list) reproduces the `MessageOrderError` path instead.

</details>

#### A regression test does not actually need external SSH implementations

The earlier discussion on #2083 raised concerns about test feasibility ("most SSH implementations work differently"). That concern is valid for *interoperability* tests, but the stale-snapshot bug itself can be pinned with a paramiko-only test using `LoopSocket` and the existing `TransportTest` setUp. The minimal assertion is: with `_modulus_pack is None`, a handshake should still succeed when the client offers GEX + a non-GEX fallback.

<details>
<summary>Drop-in regression test for tests/test_transport.py</summary>

```python
def test_gex_not_advertised_when_moduli_unavailable(self):
    """
    Regression test for #2126 / #2083.

    When the server's ``_modulus_pack`` is None (i.e. ``load_server_moduli()``
    was not called or returned no usable groups), the server must not
    advertise ``diffie-hellman-group-exchange-sha*`` in its KEXINIT.

    Before the fix in #2083, ``_send_kex_init()`` filtered
    ``security_options.kex`` but left a stale local ``kex_algos``
    still containing GEX, which was sent to the client. The exact
    failure mode depends on the client's kex preferences:

    - GEX-only client: server's ``_parse_kex_init`` intersection
      is empty, raising ``IncompatiblePeer`` (paramiko >= 5.0)
      or ``MessageOrderError`` (paramiko 4.x).
    - GEX + non-GEX fallback (this test's scenario): the server
      agrees on the non-GEX fallback, but the client selects GEX
      from the server's stale KEXINIT; the packet-type mismatch
      then raises ``MessageOrderError`` on both 4.x and 5.x.

    With the fix, the server's KEXINIT no longer advertises GEX,
    so the client falls back to a non-GEX kex algorithm and the
    handshake succeeds.
    """
    self.assertIsNone(self.ts._modulus_pack)

    # Client prefers GEX (sha256) but accepts group14-sha256 as fallback.
    # (paramiko 5.0 no longer ships the SHA-1 GEX variant, so SHA-256 is
    # the only GEX option that exists in the runtime kex registry on >= 5.0.)
    self.tc.get_security_options().kex = (
        "diffie-hellman-group-exchange-sha256",
        "diffie-hellman-group14-sha256",
    )

    host_key = RSAKey.from_private_key_file(_support("rsa.key"))
    public_host_key = RSAKey(data=host_key.asbytes())
    self.ts.add_server_key(host_key)
    event = threading.Event()
    server = NullServer()
    self.ts.start_server(event, server)
    self.tc.connect(
        hostkey=public_host_key,
        username="slowdive",
        password="pygmalion",
    )
    event.wait(1.0)

    self.assertTrue(self.ts.is_active(),
        "server transport should be active after non-GEX fallback")
    self.assertTrue(self.tc.is_active(),
        "client transport should be active after non-GEX fallback")
```

Notes:
- Reuses the existing `TransportTest` infrastructure (`LoopSocket`, `NullServer`, `_support("rsa.key")`).
- The `_modulus_pack is None` precondition is the bug trigger; we assert it explicitly so the test stays meaningful if the test-suite setup ever changes.
- The kex preference `(GEX-sha256, group14-sha256)` makes the client *want* GEX while still allowing a clean fallback once the server is fixed.
- We deliberately do **not** assert on `self.tc.kex_engine.__class__` after the handshake — paramiko clears `self.kex_engine = None` inside [`_parse_newkeys()` (transport.py L2747)](https://github.com/paramiko/paramiko/blob/5.0.0/paramiko/transport.py#L2747), so any post-handshake assertion on the engine type would pass vacuously.

</details>

#### Workaround (still effective on 5.0.0)

For downstream users hitting this until #2083 lands, the `disabled_algorithms` workaround still works on paramiko 5.0.0:

```python
session = paramiko.Transport(client_sock)
if not paramiko.Transport.load_server_moduli():
    # (paramiko 5.0 no longer ships the SHA-1 GEX variant, so SHA-256 alone is sufficient.)
    session.disabled_algorithms = {
        "kex": ["diffie-hellman-group-exchange-sha256"]
    }
```

Thanks to @mdietzer-fn for the original report and to @manfred-kaiser for the fix in #2083 — hope this paramiko 5.0 evidence is useful.

---
_AI assisted (Claude Code, Opus 4.7), verified locally on paramiko 5.0.0._

**Comment by Route-Reflector** [comment]
### Follow-up: the `#2126` workaround hits an empty-overlap KEX failure on Win/macOS

Follow-up note from a downstream project after deploying the workaround discussed in this issue. On `paramiko 5.0.0` the server-mode CI matrix failed on Win/macOS while Linux passed.

**TL;DR:**
- The `#2126` `disabled_algorithms` workaround only fires when moduli is missing — it's a no-op on hosts that have moduli (e.g. Linux). On hosts without moduli (e.g. our Win/macOS CI), dropping GEX leaves no common KEX with some legacy clients on `paramiko 5.0.0`.
- Root cause is the intersection of two individually correct decisions: the workaround drops GEX when moduli is missing, and we don't see any SHA-1 KEX in the current `_preferred_kex` (verified on `5.0.0`). Together they leave zero overlap with legacy clients whose remaining non-SHA-1 overlap depended on GEX.
- `#2083` fixes the `#2126` bug itself (the stale-snapshot crash) and makes the workaround redundant — worth landing. The empty-overlap symptom we hit here is a separate layer; that side remains.

#### Why Linux passes and Win/macOS fails

The OS dependence is driven entirely by whether `paramiko.Transport.load_server_moduli()` returns True. The workaround, applied before `start_server()`:

```python
session = paramiko.Transport(sock)
if not paramiko.Transport.load_server_moduli():
    session.disabled_algorithms = {
        "kex": ["diffie-hellman-group-exchange-sha256"]
    }
# ... then session.start_server(...)
```

(On `5.0.0`, disabling GEX-SHA256 alone is sufficient.)

| Runner OS | `/etc/ssh/moduli` | `load_server_moduli()` | Workaround fires? | Server advertises GEX? (after workaround) |
| --- | --- | --- | --- | --- |
| Linux | yes (OpenSSH default) | yes | **no** | yes |
| Windows | no | no | **yes** | no |
| macOS | no in standard locations on our runner | no | **yes** | no |

On Linux the workaround simply doesn't apply, so GEX stays advertised and legacy clients still find an overlap. On Win/macOS the workaround correctly fires for `#2126`, GEX gets dropped, and those clients have nothing left to negotiate against.

#### How the intersection goes to zero on moduli-less hosts

The current `_preferred_kex` (observed on `5.0.0`) doesn't contain any SHA-1 KEX — the right call security-wise. After the `#2126` workaround drops the GEX-SHA256 entry on top of that already-tightened list, the intersection with legacy clients whose remaining non-SHA-1 overlap was GEX-SHA256 (older network gear, `netmiko`'s legacy device classes, etc.) hits zero.

To be clear about layering: without the `#2126` workaround these same clients would also fail on `paramiko 5.0.0` — but via the `#2126` stale-snapshot crash, not via empty-overlap. The workaround turns a crash into a clean negotiation failure; the empty-overlap is the residual symptom.

<details>
<summary>Verified <code>_preferred_kex</code> on paramiko 5.0.0 + a concrete empty-overlap example</summary>

`_preferred_kex` is a tuple — preference order matters, so showing it preserved (with `KexCurve25519.is_available() == True`, which is the common path):

```python
>>> paramiko.Transport._preferred_kex
(
    "curve25519-sha256@libssh.org",
    "ecdh-sha2-nistp256",
    "ecdh-sha2-nistp384",
    "ecdh-sha2-nistp521",
    "diffie-hellman-group16-sha512",
    "diffie-hellman-group-exchange-sha256",
    "diffie-hellman-group14-sha256",
)
```

After the `#2126` workaround removes the GEX-SHA256 entry on a host without moduli, the server advertises (preference order preserved):

```
curve25519-sha256@libssh.org
ecdh-sha2-nistp256
ecdh-sha2-nistp384
ecdh-sha2-nistp521
diffie-hellman-group16-sha512
diffie-hellman-group14-sha256
```

A legacy client preference (concrete example, `netmiko.fortinet.FortinetSSH.preferred_kex` — GEX-SHA256 plus SHA-1 fallbacks):

```python
{
    "diffie-hellman-group14-sha1",
    "diffie-hellman-group-exchange-sha1",
    "diffie-hellman-group-exchange-sha256",
    "diffie-hellman-group1-sha1",
}
```

Set intersection is empty → KEX negotiation fails → handshake never completes. On `paramiko 4.x` the same intersection still contained `diffie-hellman-group14-sha1`, which is why this case was silent until now.

</details>

#### How `#2083` helps (and where it stops)

`#2083` fixes the underlying `#2126` bug — paramiko itself produces a coherent KEXINIT that omits GEX when moduli is missing, making the downstream `disabled_algorithms` workaround **redundant**. That's the right fix and worth landing.

The empty-overlap symptom is a separate layer though: `#2083` is about *whether the KEXINIT is internally consistent*, not about *what KEX is offered at all* on moduli-less hosts. After `#2083` the server still ends up GEX-less on those hosts (just produced cleanly inside paramiko), so the no-common-kex case with legacy clients persists.

#### Downstream workaround (practical)

If you need GEX on hosts without a system moduli file (common on Win/macOS CI), bundle a `moduli` file and call `paramiko.Transport.load_server_moduli(filename=...)` before `start_server()`. This restores GEX advertisement, but the downstream then owns maintaining that moduli file over time (updates / regeneration / policy decisions).

Thanks again to the maintainers for the `5.0` hardening work — flagging this only so other downstreams don't trip on the same edge.

---
_AI assisted (Claude Code, Opus 4.7), verified locally on paramiko 5.0.0._
